probe-rs/examples/benchmark.rs

```diff
@@ -0,0 +1,244 @@
+
+    let protocol_speed = if let Some(speed) = matches.speed {
+        let protocol_speed = probe
+            .set_speed(speed)
+            .map_err(|_| "Failed to set probe speed")?;
+        protocol_speed
+    } else {
+        let protocol_speed = probe
+            .set_speed(10000)
+            .map_err(|_| "Failed to set probe speed")?;
+        protocol_speed
+    } as i32;
+
+    if ![100, 1000, 10000, 50000].contains(&protocol_speed) {
+        return Err("Speed must be in [100, 1000, 10000, 50000] KHz");
+    }
+
+    let probe_name = probe.get_name();
+
+    let mut session = probe
+        .attach(target_selector)
+        .map_err(|_| "Failed to attach probe to target")?;
+    let mut core = session.core(0).map_err(|_| "Failed to attach to core")?;
+
+    let data_size_words = SIZE;
+
+    let data_size_bytes = data_size_words * 4;
+
+    let mut rng = rand::thread_rng();
+
+    let mut sample_data = vec![0u32; data_size_words];
+
+    rng.fill(&mut sample_data[..]);
+
+    core.halt(Duration::from_millis(100))
+        .expect("Halting failed");
+
+    let write_start = Instant::now();
+    core.write_32(matches.address, &sample_data)
+        .expect("Writing the sample data failed");
+
+    let write_duration = write_start.elapsed();
+
+    let write_throughput = (data_size_bytes as f32) / write_duration.as_secs_f32();
+
+    println!(
+        "Wrote {} bytes in {:?} ({:>8.2} bytes/s)",
+        data_size_words * 4,
+        write_duration,
+        write_throughput
+    );
+
+    // read back data
+
+    let mut readback_data = vec![0u32; data_size_words];
+
+    let read_start = Instant::now();
+    core.read_32(matches.address, &mut readback_data)
+        .expect("Reading the sample data failed");
+    let read_duration = read_start.elapsed();
+
+    let read_throughput = (data_size_bytes as f32) / read_duration.as_secs_f32();
+
+    println!(
+        "Read  {} bytes in {:?} ({:>8.2} bytes/s)",
+        data_size_words * 4,
+        read_duration,
+        read_throughput
+    );
+
+    if sample_data != readback_data {
+        let mismatch = sample_data
+            .iter()
+            .zip(readback_data.iter())
+            .position(|(sample, readback)| sample != readback);
+
+        eprintln!("Verification failed!");
+
+        if let Some(mismatch) = mismatch {
+            eprintln!(
+                "Readback data differs at address {:08x}: expected word {:08x}, got word {:08x}",
+                matches.address, sample_data[mismatch], readback_data[mismatch]
+            );
+        }
+    } else {
+        println!("Verification succesful.");
+        let start = SystemTime::now();
+        let since_the_epoch = start
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs();
+
+        let commit_hash = String::from_utf8_lossy(
+            &Command::new("git")
+                .args(&["rev-parse", "--short", "HEAD"])
```

Comment:
This doesn't detect if there are local modifications. We should probably detect that, to avoid submitting benchmarks with local modifications by accident.

Reply:
Is this solved?

Reply:
Yes, here: https://github.com/probe-rs/probe-rs/pull/514/files#diff-9ac4fa0d7ef685c416b61409147de558c37209d8c247832cdfaab49ce475352fR159